[Writing Tools] Support semantic HTML lists as input to Writing Tools

rr-codes opened the pull request:
#### 86b2ec59d30c53e20e5ba5fbe434a2a2b0e3004a
<pre>
[Writing Tools] Support semantic HTML lists as input to Writing Tools
<a href="https://bugs.webkit.org/show_bug.cgi?id=296301">https://bugs.webkit.org/show_bug.cgi?id=296301</a>
<a href="https://rdar.apple.com/146968196">rdar://146968196</a>

Reviewed by Wenson Hsieh.

When gathering the text context for Writing Tools, the resulting NSAttributedString needs
to have its relevant paragraph styles marked with NSTextLists so that Writing Tools can understand
that the content has lists in it and can preserve them.

To facilitate this, a new option is added to EditingHTMLConverter which enables extracting NSTextLists.
This is done by recursively traversing upwards from each element to find the containing list elements,
and then convert them to NSTextLists. The algorithm also employs caching and memoization to optimize
the amount of traversals needed.

* Source/WebCore/editing/cocoa/EditingHTMLConverter.h:

Add a `TextLists` option so that clients can opt-in to this new behavior.

* Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
(WebCore::enclosingElement):
(WebCore::enclosingLinkElement):
(WebCore::enclosingListElement):

Generalize the `enclosingLinkElement` function so that it can be reused for any arbitrary predicate,
in this case for determining if something is a list element or not.

(WebCore::associateElementWithTextLists):

The primary function used to do the extraction using a recursive algorithm; the output of the algorithm
is mutations to the `textListsForListElements` parameter. More details can be found in the inline comments.

(WebCore::updateAttributes):
(WebCore::editingAttributedStringInternal):

Modify the paragraph style of the attributes to account for the NSTextLists.

* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::willBeginWritingToolsSession):

Use the `TextLists` option.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(htmlToAttributedString):
(runTest):
(TEST(WritingToolsContextGeneration, ContextWithBasicOrderedList)):
(ContextWithCustomizedOrderedList)):
((WritingToolsContextGeneration, ContextWithNestedLists)):
((WritingToolsContextGeneration, ContextWithDiscontiguousLists)):
((WritingToolsContextGeneration, ContextWithBasicUnorderedList)):
((WritingToolsContextGeneration, ContextWithStyledContentChildrenInList)):

Add a lot of tests in a unique test suite for validation of the algorithm for different content.

* Tools/TestWebKitAPI/cocoa/DecomposedAttributedText.h: Added.
(_DecomposedAttributedTextBold::_DecomposedAttributedTextBold):
(_DecomposedAttributedTextItalic::_DecomposedAttributedTextItalic):
(_DecomposedAttributedTextOrderedList::_DecomposedAttributedTextOrderedList):
(_DecomposedAttributedTextUnorderedList::_DecomposedAttributedTextUnorderedList):
* Tools/TestWebKitAPI/cocoa/DecomposedAttributedText.mm: Added.
(convertMarker):
(decomposeChildren):
(decomposeAttributedTextRuns):
(runsForAttributedText):
(decompose):
(operator==):
(operator&lt;&lt;):

Add a testing mechanism to make it easier to compare HTML string converted to an NSAttributedString
value, and some expected value expressed as an HTML-like structure DSL.

Canonical link: <a href="https://commits.webkit.org/297769@main">https://commits.webkit.org/297769@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b3351e86b957c076dd367bd03d56f20dfd65b1d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112718 "12 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32450 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22928 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118917 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63214 "Built successfully") | [✅ 🛠 ios-apple](https://apple.com) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114680 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41013 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85792 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36430 "") | [⏳ 🛠 mac-apple ](url) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115665 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26414 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101402 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66094 "Found 141 new API test failures: /WPE/TestLoaderClient:/webkit/WebKitWebView/loading-status, /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification, /WPE/TestWebKitFindController:/webkit/WebKitFindController/options, /WPE/TestLoaderClient:/webkit/WebKitWebView/reload, /WPE/TestWebKitUserContentManager:/webkit/WebKitUserContentManager/script-message-received, /WPE/TestWebKitWebView:/webkit/WebKitWebView/run-javascript, /WPE/TestWebKitWebView:/webkit/WebKitWebView/disable-web-security, /WPE/TestLoaderClient:/webkit/WebKitWebView/load-plain-text, /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification-initial-permission-allowed, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme ... (failure)") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25719 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19532 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62676 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95813 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19609 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122138 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39792 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29665 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94657 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40175 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97635 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94394 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24109 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39509 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17319 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35895 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39680 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45181 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39319 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42652 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41058 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
```diff
@@ -265,8 +267,66 @@ static bool hasAncestorQualifyingForWritingToolsPreservation(Element* ancestor,
+    RetainPtr nsTextList = list.createTextList();
+    RetainPtr allLists = [ancestors arrayByAddingObject:nsTextList.get()];
+    textListsForListElements.set(*element, allLists);
+};
```
```suggestion
}
```

lilyspiniolas commented on Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
```diff
@@ -362,13 +422,33 @@ static void updateAttributes(const Node* node, const RenderStyle& style, OptionS
+
+    if (includedElements.contains(IncludedElement::TextLists)) {
+        if (RefPtr enclosingList = enclosingListElement(*node, enclosingListCache)) {
+            RetainPtr<NSMutableParagraphStyle> paragraphStyle = adoptNS([(attributes[NSParagraphStyleAttributeName] ?: [PlatformNSParagraphStyle defaultParagraphStyle]) mutableCopy]);
```
nit: this is entirely opinion-based but I think it might look a little cleaner to have everything inside `adoptNS` on a separate line

lilyspiniolas replied:
not the whole thing, but just pulling the `attributes[NSParagraphStyleAttributeName]` check outside, if possible

whsieh commented on Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
```diff
@@ -265,8 +267,66 @@ static bool hasAncestorQualifyingForWritingToolsPreservation(Element* ancestor,
+
+    // Base case #2:
+    // If the element is already associated with an array of text lists, no work needs to be done.
+    if (textListsForListElements.find(*element) != textListsForListElements.end())
```
A bit cleaner, since we're not using the result:

```suggestion
    if (textListsForListElements.contains(*element))
```

whsieh commented on Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
```diff
@@ -265,8 +267,66 @@ static bool hasAncestorQualifyingForWritingToolsPreservation(Element* ancestor,
+    }
+
+    TextList list;
+    list.styleType = element->renderer()->style().listStyleType();
```
Is the element guaranteed to have a renderer here?

whsieh commented on Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
```diff
@@ -362,13 +422,33 @@ static void updateAttributes(const Node* node, const RenderStyle& style, OptionS
+                [paragraphStyle setTextLists:@[ ]];
+                attributes[NSParagraphStyleAttributeName] = paragraphStyle.get();
+            }
+        }
```
```suggestion
        } else if (RetainPtr<NSMutableParagraphStyle> paragraphStyle = adoptNS([attributes[NSParagraphStyleAttributeName] mutableCopy])) {
            [paragraphStyle setTextLists:@[ ]];
            attributes[NSParagraphStyleAttributeName] = paragraphStyle.get();
        }
```

rr-codes committed:
[Writing Tools] Support semantic HTML lists as input to Writing Tools
https://bugs.webkit.org/show_bug.cgi?id=296301
rdar://146968196

Reviewed by Wenson Hsieh.

When gathering the text context for Writing Tools, the resulting NSAttributedString needs
to have its relevant paragraph styles marked with NSTextLists so that Writing Tools can understand
that the content has lists in it and can preserve them.

To facilitate this, a new option is added to EditingHTMLConverter which enables extracting NSTextLists.
This is done by recursively traversing upwards from each element to find the containing list elements,
and then convert them to NSTextLists. The algorithm also employs caching and memoization to optimize
the amount of traversals needed.

* Source/WebCore/editing/cocoa/EditingHTMLConverter.h:

Add a `TextLists` option so that clients can opt-in to this new behavior.

* Source/WebCore/editing/cocoa/EditingHTMLConverter.mm:
(WebCore::enclosingElement):
(WebCore::enclosingLinkElement):
(WebCore::enclosingListElement):

Generalize the `enclosingLinkElement` function so that it can be reused for any arbitrary predicate,
in this case for determining if something is a list element or not.

(WebCore::associateElementWithTextLists):

The primary function used to do the extraction using a recursive algorithm; the output of the algorithm
is mutations to the `textListsForListElements` parameter. More details can be found in the inline comments.

(WebCore::updateAttributes):
(WebCore::editingAttributedStringInternal):

Modify the paragraph style of the attributes to account for the NSTextLists.

* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::willBeginWritingToolsSession):

Use the `TextLists` option.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
(htmlToAttributedString):
(runTest):
(TEST(WritingToolsContextGeneration, ContextWithBasicOrderedList)):
(ContextWithCustomizedOrderedList)):
((WritingToolsContextGeneration, ContextWithNestedLists)):
((WritingToolsContextGeneration, ContextWithDiscontiguousLists)):
((WritingToolsContextGeneration, ContextWithBasicUnorderedList)):
((WritingToolsContextGeneration, ContextWithStyledContentChildrenInList)):

Add a lot of tests in a unique test suite for validation of the algorithm for different content.

* Tools/TestWebKitAPI/cocoa/DecomposedAttributedText.h: Added.
(_DecomposedAttributedTextBold::_DecomposedAttributedTextBold):
(_DecomposedAttributedTextItalic::_DecomposedAttributedTextItalic):
(_DecomposedAttributedTextOrderedList::_DecomposedAttributedTextOrderedList):
(_DecomposedAttributedTextUnorderedList::_DecomposedAttributedTextUnorderedList):
* Tools/TestWebKitAPI/cocoa/DecomposedAttributedText.mm: Added.
(convertMarker):
(decomposeChildren):
(decomposeAttributedTextRuns):
(runsForAttributedText):
(decompose):
(operator==):
(operator<<):

Add a testing mechanism to make it easier to compare HTML string converted to an NSAttributedString
value, and some expected value expressed as an HTML-like structure DSL.

Canonical link: https://commits.webkit.org/297769@main
